crates/bevy_pbr/src/meshlet/mod.rs

```diff
@@ -0,0 +1,293 @@
+    core_3d::{graph::node::*, Camera3d, CORE_3D},
+    prepass::{DeferredPrepass, MotionVectorPrepass, NormalPrepass},
+};
+use bevy_ecs::{
+    bundle::Bundle,
+    entity::Entity,
+    query::Has,
+    schedule::IntoSystemConfigs,
+    system::{Commands, Query},
+};
+use bevy_render::{
+    render_graph::{RenderGraphApp, ViewNodeRunner},
+    render_resource::{Shader, TextureUsages},
+    renderer::RenderDevice,
+    settings::WgpuFeatures,
+    view::{prepare_view_targets, InheritedVisibility, Msaa, ViewVisibility, Visibility},
+    ExtractSchedule, Render, RenderApp, RenderSet,
+};
+use bevy_transform::components::{GlobalTransform, Transform};
+use bevy_utils::tracing::warn;
+
+const MESHLET_BINDINGS_SHADER_HANDLE: Handle<Shader> = Handle::weak_from_u128(1325134235233421);
+const MESHLET_VISIBILITY_BUFFER_RESOLVE_SHADER_HANDLE: Handle<Shader> =
+    Handle::weak_from_u128(2325134235233421);
+const MESHLET_MESH_MATERIAL_SHADER_HANDLE: Handle<Shader> =
+    Handle::weak_from_u128(3325134235233421);
+
+/// Provides a plugin for rendering high-poly 3d meshes using an efficient GPU-driven method. See also [`MeshletMesh`].
+///
+/// Rendering high-poly meshes with thousands or millions of triangles is extremely expensive in Bevy's standard renderer.
+/// Once pre-processed into a [`MeshletMesh`], this plugin can render these kinds of meshes very efficently via the following method:
+/// * All work is done on the GPU. Minimal CPU processing is required, unlike Bevy's standard renderer.
+/// * Individual meshlets outside of the camera's frustum are culled (unlike Bevy's standard renderer, which can only cull entire meshes).
+/// * All meshlets that were visible last frame (and are in the camea's frustum this frame) get rendered to a depth buffer.
+/// * The depth buffer is then downsampled to form a hierarchical depth buffer.
+/// * All meshlets that were not _not_ visible last frame get frustum culled and tested against the depth buffer, and culled if they would not be visible (occlusion culling, which Bevy's standard renderer does not have).
+/// * A visibility buffer is then rendered for the surviving frustum and occlusion culled meshlets. Each pixel of the texture encodes the visible meshlet and triangle ID.
+/// * For the opaque and prepass phases, one draw per [`Material`] batch (unique pipeline + bind group) is performed, regardless of the amount of entities using that material.
+///   The material's fragment shader reads the meshlet and triangle IDs from the visibility buffer to reconstruct the rendered point on the mesh and shade it.
+///
+/// This plugin is not compatible with [`Msaa`], and adding this plugin will disable it.
+///
+/// This plugin requires support for [push constants](WgpuFeatures#associatedconstant.PUSH_CONSTANTS), and will not work on web platforms.
+///
+/// ![A render of the Stanford dragon as a `MeshletMesh`](https://raw.githubusercontent.com/bevyengine/bevy/meshlet/crates/bevy_pbr/src/meshlet/meshlet_preview.png)
```

Comment:
I think that this would be a first for the repo. Has this (or this sort of thing) been discussed somewhere?

Reply:
We do have some images in https://github.com/bevyengine/bevy/blob/main/docs/profiling.md, but I think they were just uploaded in a GH issue or something?

Reply:
@alice-i-cecile @mockersf any thoughts?

Reply:
not a first: https://github.com/bevyengine/bevy/blob/c9e1fcdb355b049fa3c3df8cb1cd1f4343f1b9d1/crates/bevy_render/src/mesh/mesh/mod.rs#L82

But I think the image shouldn't be in the crate, the asset folder would have my preference.

Reply:
Oh, okay, cool.

I'd love to see the window decorations cropped out and the image resized a bit. It seems excessively large, resulting in a file size of ~900kb. On my high dpi display, it could be resized to ~42%.

The contents of the desktop behind the window are also visible on the left and right edges.

I recall some sort of guidance from Cart about blog post images not including decorations just for a consistent look/feel, which feels like it could be good to follow here as well.

Reply:
I was going to redo `meshlet_preview.png` for you but this is what the example looks like right now on my m1 mac:
<img width="1280" alt="image" src="https://github.com/bevyengine/bevy/assets/200550/91e54a86-57ca-404f-97ec-95c9f9aa5837">

Reply:
May have been a driver issue. Things seem okay after upgrading to MacOS 14.2 (from 13.6).

Reply:
Took a new screenshot with the dragon on Bevy's fresh new on-brand clear color. Just dropped the system from the screenshot example in, if you're looking for a way to get clean screenshots.

![meshlet_preview_dragon](https://github.com/bevyengine/bevy/assets/200550/efd41cc7-5b2d-4992-a5c4-4fdbec3cba68)

This is still ~300k which doesn't seem great. But we don't really have any guidelines about this sort of thing, so 🤷.

Reply:
Fresh dragon on Bevy's new on-brand clear color. Still ~300k at this size, which doesn't seem ideal. IMO I don't know if we need this image at all, but this one's a bit nicer if we want it.

![screenshot-0](https://github.com/bevyengine/bevy/assets/200550/ed1865e1-1405-4fa8-bd58-60db7d97398f)